feature(yum_versionlock): add support to pin specific package versions

gyptazy opened the pull request:
feature(yum_versionlock): add support to pin specific package versions

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Adjusted `yum_versionlock` (as requested in #4470) to handle version suffixes in a correct way to avoid wrong change results. This will validate and handle the following cases:
  - package name like:             httpd
  - package name and version like: httpd-0:2.4.57-3.el9

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #4470
<!--- Please do not forget to include a changelog fragment:
      https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-changelog-fragments
      No need to include one for docs-only or test-only PR, and for new plugin/module PRs.
      Read about more details in CONTRIBUTING.md.
      -->

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
yum_versionlock

##### ADDITIONAL INFORMATION
@tedgin & @matt-horwood-mayden as you requested this change, please feel free to test.



<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
Before:
```
[gyptazy@localhost ansible]# ansible-playbook playbooks/yum_versionlock.yml 

PLAY [yum versionlock] ***************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [lock iCommands to version 4.2.8] ***********************************************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

After:
```
[gyptazy@localhost ansible]# ansible-playbook playbooks/yum_versionlock.yml 

PLAY [yum versionlock] ***************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [lock iCommands to version 4.2.8] ***********************************************************************************************************************************************************************************************************
ok: [localhost]

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

matt-horwood-mayden commented:
can confirm this is fixed, please merge

matt-horwood-mayden commented:
@BenMayden & @ HiredMussel

gyptazy commented:
> can confirm this is fixed, please merge

Thanks for testing!
Ping @felixfontein

felixfontein commented on plugins/modules/yum_versionlock.py:
```diff
@@ -122,7 +121,15 @@ def match(entry, name):
+    if name in m.group():
+        return True
+    else:
+        return False
```
1. Why not simply `return name in m.group()`? That's equivalent to the above four lines.
2. If the package name is `ttp`, it will also match `httpd`. That looks wrong to me.
3. I think `fnmatch` is used to also match wildcards (see the documentation for `name`, it explicitly mentions them). The new code doesn't handle wildcards.

gyptazy replied:
Ok, we can keep the fnmatch solution. Will push later.

felixfontein reviewed (commented):
Thanks for your contribution! The fix doesn't look correct to me, see my comments below.

Please also note that PRs generally need [changelog fragments](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment) (except in a few special cases, like when adding new plugins/modules, or for docs-only PRs).

gyptazy commented:
Hey @felixfontein,

thanks for the hint regarding the changeling fragments. The requested changes have been added.

Cheers,
gyptazy

felixfontein commented on changelogs/fragments/6861-yum_versionlock-fix-packages-with-versions.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - yum_versionlock - Fix package handling with specific versions. (https://github.com/ansible-collections/community.general/pull/6861, https://github.com/ansible-collections/community.general/issues/4470).
```
```suggestion
  - yum_versionlock - fix package handling with specific versions (https://github.com/ansible-collections/community.general/pull/6861, https://github.com/ansible-collections/community.general/issues/4470).
```

felixfontein commented on plugins/modules/yum_versionlock.py:
```diff
@@ -111,18 +111,26 @@ def get_versionlock_packages(self):
-    return fnmatch(m.group("name"), name)
+    if fnmatch(m.group("name"), name):
+        match = True
+    if fnmatch(m.group(), '{0}*'.format(name)):
```
Doesn't this mean that if I ask to lock a package called `foo`, that it will also match a package called `foobar`? Using `'{0}-*` has a similar problem, `foo` will also match `foo-bar`. (Assuming `-` is allowed in package names. I don't use yum/dnf based systems, so :shrug:.)

Also your fix only seems to work with yum, but not with dnf, according to https://github.com/ansible-collections/community.general/blob/b9b0ea855abbd1c239688f582caafd95e82b48c5/plugins/modules/yum_versionlock.py#L90-L93.

gyptazy replied:
I tested this on CentOS 7 with legacy yum as well as with CentOS Stream where yum is just an alias for dnf.

felixfontein commented:
As mentioned in https://github.com/ansible-collections/community.general/issues/4470#issuecomment-1626033246: this should not be a bugfix, but a new feature. The module never explicitly supported providing a version.

(It might make sense to pass the version in as a separate parameter, so that the module doesn't have to try to parse the version from `name`. I have no idea how easy/hard it is to split off versions and to detect whether they are present in the first place.)

gyptazy commented:
@felixfontein, should these adjustments be done within this PR (by editing the type) or a new one (might be misleading bugfix/new feature)?

gyptazy commented:
Okay, here we go @felixfontein.

1. Adjusted PR content (let's just validate the output from yum versionlock and compare it with the given package name(s) with optional versions passed to Ansible. We do not need to split it here to remove the version because that's also the part we pass to Ansible in that case we want to pin a specific version).
2. Adjusted PR and meta infos to `Feature Pull Request` 
3. Adjusted docs and example(s)
4. Adjusted changelog fragments
5. Tested with all constellation with:
- httpd-devel-0:2.4.57-2.el9.*
- lighttpd-0:1.4.67-1.el9.*
- httpd-0:2.4.57-2.el9.*
- httpd-devel
- httpd
- lighttpd

Hope this fits your needs, now.

Cheers,
gyptazy

felixfontein commented on plugins/modules/yum_versionlock.py:
```diff
@@ -25,7 +25,7 @@
   name:
     description:
-      - Package name or a list of package names with optional wildcards.
+      - Package name or a list of package names with optional version or wildcards.
```
Let's be explicit about since when this is supported:
```suggestion
      - Package name or a list of package names with optional version or wildcards.
      - Specifying versions is supported since community.general 7.2.0.
```

gyptazy replied:
ok, added.

felixfontein commented on plugins/modules/yum_versionlock.py:
```diff
@@ -111,18 +118,26 @@ def get_versionlock_packages(self):
-    return fnmatch(m.group("name"), name)
+    if fnmatch(m.group("name"), name):
+        match = True
+    if entry.rstrip('.*') == name:
```
I'm not sure whether this does what you expect. It removes all trailing characters from `entry` that are `.` or `*`. So `foo****...***.*` will be converted to `foo`. But `foo.123` will not be changed.

gyptazy replied:
So, this is just rstrip, not regex.
In case we have (return output from yum_versionlock list):
`httpd-devel-0:2.4.57-2.el9.*`
it will be
`httpd-devel-0:2.4.57-2.el9`. Now, we can compare it with the name passed to Ansible.

gyptazy replied:
@felixfontein ping

felixfontein commented:
As far as I can judge this looks good. Will merge once CI passes (ignoring that ansible/third-party-check one that always fails at the moment :) ).

felixfontein commented:
@gyptazy thanks for fixing this!
@matt-horwood-mayden thanks for testing!
